### DISCUSSION TITLE
JPH::Ref casts

### DISCUSSION THREAD

**Discussion by fhoenig** [jrouwe/JoltPhysics#1132] [General]
What's the suggested idiom for down casting and type checking for objects held by JPH::Ref ?
I personally mostly implement "is<T>" "as<T>".

**Comment by jrouwe** [comment]
Jolt has the `StaticCast` template, but it comes from my old engine which predates C++11. It is barely used and thus broken. I've created a fix in #1135.

Alternatively you can use:

```
static_cast<DowncastType *>(my_ref.GetPtr())
```

**Reply by fhoenig** [reply]
Ah okay. I thought there was maybe a shortcut somewhere.

```
if (my_ref.is<ConeConstraint>())
{
    auto& c = my_ref.as<ConeConstraint>()
}
```

But then again, its a joy how fast jolt compiles currently :-)

**Reply by jrouwe** [reply]
You could write:

```
if (MyClass *c = DynamicCast<MyClass>(my_ref))
{
   ...
}
```

but unfortunately it won't work for `Constraint` as it doesn't have RTTI so you need to do:

```
if (my_ref->GetSubType() == EConstraintSubType::Cone)
{
   ConeConstraint *c = StaticCast<ConeConstraint>(my_ref);
}
```

The reason why `Constraint` doesn't have RTTI is that for run-time classes (e.g. Shape, Constraint) I was planning to move towards classes that can just be blitted into memory. So, instead of a vfptr they get a type enum.

**Reply by fhoenig** [reply]
Yeah the 3rd is what I wrote everywhere.  Also +1 one on moving to more plain data structures.
